src/components/selectable-table/SelectableTable.js

```diff
@@ -0,0 +1,155 @@
+import union from 'lodash/union';
+import difference from 'lodash/difference';
+import isEqual from 'lodash/isEqual';
+
+/**
+ * A table with "selectable" rows.
+ * @prop {string[]} columns An array of column names, corresponding to data object properties.
+ * @prop {object[]} data An array of data objects used to populate table rows.
+ * @prop {function} onChange Callback function,
+ * passed a selection object when `allowMultiple` is false (and `null` if `allowUncheck` is true),
+ * or passed an array of selection objects when `allowMultiple` is true.
+ * @prop {string} idKey The key for a unique identifier property of `data` objects.
+ * @prop {string} valueKey If initially-selected rows are required,
+ * this key specifies a boolean property of the `data` objects
+ * indicating those rows that should be initially selected.
+ * @prop {boolean} allowMultiple Whether to allow multiple rows to be selected. By default, false.
+ * @prop {boolean} allowUncheck Whether to allow selected rows to be un-checked. By default, false.
+ * @prop {boolean} showTableHead Whether to show the table header element. By default, true.
+ */
+export default function SelectableTable(props) {
+  const {
+    columns,
+    data,
+    onChange,
+    idKey = 'id',
+    valueKey = 'value',
+    allowMultiple = false,
+    allowUncheck = false,
+    showTableHead = true,
+  } = props;
+
+  const [selectedRows, setSelectedRows] = useState(null);
+
+  // Callback function to update the `selectedRows` state.
+  const onSelectRow = useCallback((value, checked) => {
+    if (!allowMultiple) {
+      if (checked) {
+        setSelectedRows([value]);
+      } else if (!checked && allowUncheck) {
+        setSelectedRows([]);
+      }
+    } else if (checked) {
+      setSelectedRows(union(selectedRows || [], [value]));
+    } else if (!checked && allowUncheck) {
+      setSelectedRows(difference(selectedRows || [], [value]));
+    }
+  }, [allowMultiple, allowUncheck, selectedRows]);
+
+  // Handler for checkbox input elements.
+  const handleInputChange = useCallback((event) => {
+    const { target } = event;
+    const { checked } = target;
+    const { value } = target;
+    onSelectRow(value, checked);
+  }, [onSelectRow]);
+
+  // Function to map row IDs to corresponding objects
+  // to pass to the `onChange` callback.
+  const getDataFromIds = useCallback(ids => ids.map(id => ({
+    [idKey]: id,
+    data: data.find(item => item[idKey] === id),
+  })), [data, idKey]);
+
+  // Function to check if a row ID has been selected.
+  const isSelected = useCallback(id => (
+    Array.isArray(selectedRows) && selectedRows.includes(id)
+  ), [selectedRows]);
+
+  /* eslint-disable react-hooks/exhaustive-deps */
+  useEffect(() => {
+    // Check whether an initial set of rows should be selected.
+    const initialSelectedRows = data
+      .map((d) => {
+        if (d[valueKey] !== undefined && d[valueKey]) {
+          return d[idKey];
+        }
+        return null;
+      })
```

Comment:
For that matter, why not:
```suggestion
      .map((d) => d[valueKey] && d[idKey])
```

Reply:
I like the if statement for readability since sometimes using the `&&` to produce some "return value" takes me a second to understand what is going on. But maybe that is just me.